core/src/main/scala/org/apache/spark/executor/Executor.scala

```diff
@@ -447,7 +450,16 @@ private[spark] class Executor(
       }
     }
 
-    val message = Heartbeat(executorId, tasksMetrics.toArray, env.blockManager.blockManagerId)
+    env.blockTransferService.getMemMetrics(this.executorMetrics)
+    val executorMetrics = if (isLocal) {
+      Utils.deserialize[ExecutorMetrics](Utils.serialize(this.executorMetrics))
```

Comment:
ah, this is a great point.  In that case, I think you can leave it in for now, but add a comment that the serialization & deserialization is just to make a copy, for to SPARK-3465